### DISCUSSION TITLE
How to publish Teal libraries to LuaRocks

### DISCUSSION THREAD

**Discussion by mtdowling** [teal-language/tl#836] [Q&A] [answered]
Let's say I'm building a library using Teal and I want to publish it to LuaRocks. From what I understand, you'd publish the generated Lua code. But what about the typing information from the original Teal code? Is there a way to publish that too, and in a way that just works well with the Teal application that's using the dependency?

**Comment by Hedwig7s** [comment]
In theory I believe you could bundle declaration files or possibly just upload both and let teal figure it out

**Reply by hishamhm** [reply, marked as answer]
Yeah, right now we don't have the most convenient story for this. I hope we can add better Teal support for LuaRocks in the near future. Right now it's really "upload both and let teal figure it out"

Doing the following in a rockspec is a bit of a hack, but it works:

```
build = {
   type = "builtin",
   modules = {
      example = "src/example.lua",
      ["example.sub"] = "src/example/sub.lua",
   },
   install = {
      lua = {
         ["example"] = "src/example.tl",
         ["example.sub"] = "src/example/sub.tl",
      },
   }
}
```

**Reply by Aire-One** [reply]
Alternatively, I use the "command" build.type to make Cyan/Teal build the Lua code automatically. 

```lua
build_dependencies = {
   "cyan",
}
build = {
   type = "command",
   build_command = "cyan build",
   install = {
      lua = {
         ["example"] = "build/example.lua",
```

`luarocks make` will build the teal to lua, then install the lua files as a rock. Only downside is that you can't ship both the lua and teal file with this method, since the build.install.lua is already populated with the built lua files.

**Comment by Aire-One** [comment]
@hishamhm, I just found https://github.com/luarocks/luarocks-build-cyan. It seems you are the author. Why isn't this rock more recommended? 

(I guess it shouldn't be that hard to add an option to include the tl files. Would you accept contribution for this feature?)

**Reply by hishamhm** [reply]
>  Why isn't this rock more recommended?

No reason in particular! I guess I just didn't do a lot of testing!

> Would you accept contribution for this feature?

Definitely!
